Добавляет рецепт git rebase --onto (#4572)
doka-guide/content: recipes/git-rebase-onto/index.md

- `to` — коммит, до которого нужно перенести изменения. Коммит, который укажете, перенесётся.

Последний аргумент можно не использовать. Тогда перенесутся все коммиты, начиная от следующего за `from` коммита до конца ветки.

Например:

```bash
git rebase --onto main feature1
```

Помните, что коммит — это всего лишь хеш? В качестве каждого аргумента можете указывать хеш, например вот так:

```bash
git rebase --onto 29aeb39e 5fd88295
```

Хеши, вообще-то, длиннее 8 символов, но git понимает короткую версию. Можете также указать полный хеш, например вот так:

```bash
git rebase --onto 29aeb39eac6b8d9e0f4cdc1459376599d1aba43c 5fd882953924b47a10794619c3063e7a50257af6
```

Вы можете указать название ветки вместо хеша, например вот так:

```bash
git rebase --onto main feature
```

Если хотите перенести несколько коммитов из основной ветки, можете использовать магию `HEAD` нотации.

```bash
git rebase --onto main HEAD^3 HEAD^
```

Эта команда перенесёт 2 коммита из текущей ветки в ветку `main`. Самый последний не перенесёт.

Рассмотрим один из примеров выше.

Было так:

![Схема двух веток Гит, бэкенд и фронтенд](images/to-rebase.png)

> Нужно вот так:

Reviewer: ```suggestion
А нужно, чтобы коммиты `Кнопочка 1` и `Тултипчик 2` оказались в ветке `backend`.
```

---
title: "Как ребейзить ветку и не думать ни о чём"

> description: "Чтобы легко управлять вашими изменениями и историей git, используйте git rebase --onto."

Reviewer 2: ```suggestion
description: "Чтобы легко управлять вашими изменениями и историей git, используйте `git rebase --onto`."
```

authors:
  - hellsquirrel
keywords:
  - git

> tags:

Reviewer 2: Нужны related

  - article
---

## Управление историей git

> Представим что у вас есть git-репозиторий с несколькими ветками. Вы работаете в одной из них и хотите перенести свои изменения в другую ветку.
> Такое приходится делать очень часто, если вы хотите актуализировать свои изменения.

Reviewer: ```suggestion
Представим что у вас есть git-репозиторий с несколькими ветками. Вы работаете в одной из них и хотите перенести свои изменения в другую ветку. Такое приходится делать очень часто, если хотите актуализировать свои изменения.
```

Рассмотрим несколько распростанëнных сценариев.

> Вы начали разрабатывать фичу и создали ветку из основной `main`, за время разработки в ветку `main` приехали изменения ваших коллег. Как переместить ваши изменения на актуальный `main`?

Reviewer: ```suggestion
Вы начали разрабатывать фичу и создали ветку из основной `main`, за время разработки в ветку `main` приехали изменения коллег. Как переместить ваши изменения на актуальный `main`?
```

```
---(Статья1)---(Статья2)---(Статья3) main
                      \
                        \---(Добавляет статью)---(Вносит правки) my-branch
```

> В вашей команде есть фронтенд и бекенд разработчики. Вы работаете в одном репозитории. Вы создали frontend-ветку из основной ветки, ваши коллега создал backend-ветку из основной. Теперь вы хотите потестировать ваш фронтенд с новым бекендом. Как это сделать?

Reviewer: ```suggestion
В вашей команде есть фронтенд- и бэкенд-разработчики. Вы работаете в одном репозитории, создали frontend-ветку из основной ветки, а ваш коллега — backend-ветку из основной. Теперь вы хотите потестировать ваш фронтенд с новым бэкендом. Как это сделать?
```

```
         /---API1---API2 backend
        /
(Статья3) main
        \
          \---(Кнопочка1)---(Тултипчик2) frontend
```

> В вашей ветке есть _несколько_ коммитов, которые нужны вашему коллеге, но кроме этого в вашей ветке есть коммиты которые сломают ветку вашего коллеги, как потестировать только не ломающие изменения?

Reviewer: ```suggestion
В вашей ветке есть _несколько_ коммитов, которые нужны коллеге, но, кроме этого, в ветке есть коммиты, которые сломают ветку коллеги. Как потестировать только не ломающие изменения?
```

```
         /---API1---API2 api
        /
(Статья3) main
        \

>           \---(Кнопочка1)---(Тултипчик2)---(😈Собака-все-ломака) my-branch

Reviewer 2: ```suggestion
          \---(Кнопочка1)---(Тултипчик2)---(😈Собака-всё-ломака) my-branch
```

```

Все эти задачи решаются с помощью команды `git rebase --onto`.

## Немного теории

> Git хранит коммиты (commits). Коммит - это просто ссылка на какое-то состояние вашего репозитория которое вы сохранили: немножко метаданных и ссылка на _предудыщий_ коммит. Сам по себе коммит это 40 символов SHA-1 хеша этого состояния.

Reviewer: ```suggestion
Git хранит коммиты (commits). Коммит — это просто ссылка на какое-то состояние репозитория, которое сохранили: немножко метаданных и ссылка на _предыдущий_ коммит. Сам по себе коммит это 40 символов SHA-1 хеша этого состояния.
```

> Ветка (branch) это ссылка на определенный коммит. У вас может быть только одна активная ветка. На нее (т. е. на опредленный коммит) указывает специальный указатель `HEAD`.

Reviewer 2: ```suggestion
Ветка (branch) это ссылка на определенный коммит. У вас может быть только одна активная ветка. На неё (т. е. на опредленный коммит) указывает специальный указатель `HEAD`.
```

> Многие команды гит принимают в качестве параметре ревизию. Так как все в гите это захешированные объекты, то в качестве ревизии мы можете передавать.

Reviewer 2: ```suggestion
Многие команды гит принимают в качестве параметре ревизию. Так как всё в гите это захешированные объекты, то в качестве ревизии мы можете передавать.
```

- Длинный хеш, например вот так `29aeb39eac6b8d9e0f4cdc1459376599d1aba43c`.
- Короткий хеш, например вот так `29aeb39e`.
- Тег, если он имеется, например вот так `v1.0.0`.
- Имя ссылки. В отсновном используется для веток. Например `main`, `origin/dev`, `refs/heads/foo`.

> Вы можете добавлять магические символы (`@{n}, ^, :`) к этим параметрам, чтобы еще эффективнее выбирать нужные коммиты. Подробности можно посмотреть [вот тут](https://git-scm.com/docs/gitrevisions). Особенно полезен символ `^` который позволяет вам выбирать предыдущие коммиты.

Reviewer: ```suggestion
Можете добавлять магические символы (`@{n}, ^, :`) к этим параметрам, чтобы ещё эффективнее выбирать нужные коммиты. Подробности можно посмотреть [в документации про `gitrevisions`](https://git-scm.com/docs/gitrevisions). Особенно полезен символ `^`. Он позволяет выбирать предыдущие коммиты.
```

## Как готовить

> Подробности можно найти например, [здесь](https://git-scm.com/docs/git-rebase).

Reviewer 2: ```suggestion
Подробности можно найти например, [в официальной документации](https://git-scm.com/docs/git-rebase).
```

Команда `git rebase` с флагом `--onto` принимает несколько аргументов:

> - `--onto` — коммит на который нужно перенести изменения. Новые коммиты добавятся сразу после него.
> - `from` — коммит с которого нужно перенести изменения Коммит который вы укажете не перенесется, а следующие за ним перенесутся.
> - `to` — коммит до которого нужно перенести изменения. Коммит который вы укажете перенесется.

Reviewer: ```suggestion
- `--onto` — коммит, на который нужно перенести изменения. Новые коммиты добавятся сразу после него.
- `from` — коммит, с которого нужно перенести изменения. Коммит, который укажете, не перенесётся, а следующие за ним перенесутся.
- `to` — коммит, до которого нужно перенести изменения. Коммит, который укажете, перенесётся.
```

> Последний аргумент можно не использовать, тогда перенесутся все коммиты начиная от следующего за `from` коммита до конца ветки.

Reviewer: ```suggestion
Последний аргумент можно не использовать. Тогда перенесутся все коммиты, начиная от следующего за `from` коммита до конца ветки.
```

Например:

```bash
git rebase --onto main feature1
```

> Помните что коммит это всего лишь хеш? В качестве каждого аргумента вы можете указывать хеш, например вот так:

Reviewer: ```suggestion
Помните, что коммит — это всего лишь хеш? В качестве каждого аргумента можете указывать хеш, например вот так:
```

```bash
git rebase --onto 29aeb39e 5fd88295
```

> Хеши вообще-то длиннее 8 символов, но git понимает короткую версию. Вы можете также указать полный хеш, например вот так:

Reviewer: ```suggestion
Хеши вообще-то длиннее 8 символов, но git понимает короткую версию. Можете также указать полный хеш, например вот так:
```

```bash
git rebase --onto 29aeb39eac6b8d9e0f4cdc1459376599d1aba43c 5fd882953924b47a10794619c3063e7a50257af6
```

Вы можете указать название ветки вместо хеша, например вот так:

```bash
git rebase --onto main feature
```

> Если вы хотите перенести несколько коммитов из основной ветки, вы можете использовать магию `HEAD` нотации.

Reviewer: ```suggestion
Если хотите перенести несколько коммитов из основной ветки, можете использовать магию `HEAD` нотации.
```

```bash
git rebase --onto main HEAD^3 HEAD^
```

> Эта команда перенесет 2 коммита из текущей ветки в ветку `main`. Самый последний не перенесет.

Reviewer 2: ```suggestion
Эта команда перенесет 2 коммита из текущей ветки в ветку `main`. Самый последний не перенесёт.
```

> Рассмотрим один из примеро выше.

Reviewer: ```suggestion
Рассмотрим один из примеров выше.
```

Было так:

```
         /---API1---API2 backend
        /
(Статья3) Ветка main
        \
          \---(Кнопочка 1)---(Тултипчик 2) frontend
```

Нужно вот так:

```
         /---API1---API2---(Кнопочка 1)---(Тултипчик 2) backend
        /
(Статья3) Ветка main
```

> Для этого нужно выполнить команду:

Reviewer: ```suggestion
Для этого выполните команду:
```

```bash
git rebase --onto backend main frontend
```

- `git rebase --onto` — переносим.
- backend — на последний коммит ветки backend, т. е. на API2.

> - main — все коммиты в моей ветке frontend, начиная с того которым заканчивается ветка main (Статья3).

Reviewer: ```suggestion
- main — все коммиты в моей ветке frontend, начиная с того, которым заканчивается ветка main (Статья3).
```

- frontend — заканчивая последним коммитом ветки frontend (Тултипчик2).

## Бонус

> Если вы хотите посмотреть какой красивой стала ваша история, используйте команду

Reviewer: ```suggestion
Если хотите посмотреть какой красивой стала ваша история, используйте эту команду:
```
